### PR TITLE
Priority Regossip

### PR DESCRIPTION
**Pull request by patrick-ogrady**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -16,7 +16,7 @@ The Subnet EVM runs in a separate process from the main AvalancheGo process and
 ```
 [v0.1.0] AvalancheGo@v1.7.0-v1.7.4
 [v0.1.1-v0.1.2] AvalancheGo@v1.7.5-v1.7.6
-[v0.2.0] AvalancheGo@v1.7.7-v1.7.8
+[v0.2.0] AvalancheGo@v1.7.7-v1.7.9
```
Comment (aaronbuchwald) [inline README.md:19]: Should this be v1.7.10 in anticipation of the next AvalancheGo release tomorrow?

File: plugin/evm/config.go [hunk 2]
```diff
@@ -24,8 +25,11 @@ const (
 	defaultMaxBlocksPerRequest                  = 0 // Default to no maximum on the number of blocks per getLogs request
 	defaultContinuousProfilerFrequency          = 15 * time.Minute
 	defaultContinuousProfilerMaxFiles           = 5
-	defaultTxRegossipFrequency                  = 1 * time.Minute
-	defaultTxRegossipMaxSize                    = 15
+	defaultTxRegossipFrequency                  = 1 * time.Second
```
Comment (aaronbuchwald) [inline plugin/evm/config.go:28]: Should this stay at 1 minute and only use 1 second as the default for priority re-gossip?

File: plugin/evm/config.go [hunk 3]
```diff
@@ -24,8 +25,11 @@ const (
...
+	defaultTxRegossipMaxSize                    = 16
+	defaultPriorityTxRegossipFrequency          = 1 * time.Second
+	defaultPriorityTxRegossipSize               = 32
+	defaultPriorityTxRegossipAddressTxs         = 16
```
Comment (aaronbuchwald) [inline plugin/evm/config.go:32]: I know these are already long, but should we make these names more specific ie. TxsPerGossipMsg, and TxsPerAddress

Comment (aaronbuchwald) [reply plugin/evm/config.go:32]: Same comment for the related names below

File: core/types/transaction.go [hunk 4]
```diff
@@ -563,7 +572,7 @@ func (t *TransactionsByPriceAndNonce) Peek() *Transaction {
 func (t *TransactionsByPriceAndNonce) Shift() {
 	acc, _ := Sender(t.signer, t.heads[0].Tx)
 	if txs, ok := t.txs[acc]; ok && len(txs) > 0 {
-		if wrapped, err := NewTxWithMinerFee(txs[0], t.baseFee); err == nil {
+		if wrapped, err := NewTxWithMinerFee(txs[0], t.baseFee, acc); err == nil {
```
Comment (aaronbuchwald) [inline core/types/transaction.go:575]: I think we can get rid of this change and use this in the push gossiper instead: https://github.com/ethereum/go-ethereum/blob/master/core/types/transaction.go#L507

File: README.md (hunk 1 above)
Comment (patrick-ogrady) [reply README.md:19]: I'll make that change when the release is ready

File: plugin/evm/config.go (hunk 2 above)
Comment (patrick-ogrady) [reply plugin/evm/config.go:28]: Good catch

File: plugin/evm/config.go [hunk 5]
```diff
@@ -90,9 +94,13 @@ type Config struct {
...
-	RemoteTxGossipOnlyEnabled bool     `json:"remote-tx-gossip-only-enabled"`
-	TxRegossipFrequency       Duration `json:"tx-regossip-frequency"`
-	TxRegossipMaxSize         int      `json:"tx-regossip-max-size"`
+	RemoteGossipOnlyEnabled       bool             `json:"remote-gossip-only-enabled"`
+	RegossipFrequency             Duration         `json:"regossip-frequency"`
+	RegossipMaxTxs                int              `json:"regossip-max-size"`
+	PriorityRegossipFrequency     Duration         `json:"tx-priority-regossip-frequency"`
+	PriorityRegossipMaxTxs        int              `json:"tx-priority-regossip-max-size"`
+	PriorityRegossipAddresses     []common.Address `json:"tx-priority-regossip-addresses"`
+	PriorityRegossipTxsPerAddress int              `json:"priority-regossip-txs-per-address"`
```
Comment (darioush) [inline plugin/evm/config.go:103]: should likely match `tx-priority-regossip-address-txs` per the docs

File: plugin/evm/config.go [hunk 6]
```diff
@@ -90,9 +94,13 @@ type Config struct {
 	KeystoreInsecureUnlockAllowed bool   `json:"keystore-insecure-unlock-allowed"`
 
 	// Gossip Settings
-	RemoteTxGossipOnlyEnabled bool     `json:"remote-tx-gossip-only-enabled"`
-	TxRegossipFrequency       Duration `json:"tx-regossip-frequency"`
-	TxRegossipMaxSize         int      `json:"tx-regossip-max-size"`
+	RemoteGossipOnlyEnabled       bool             `json:"remote-gossip-only-enabled"`
+	RegossipFrequency             Duration         `json:"regossip-frequency"`
+	RegossipMaxTxs                int              `json:"regossip-max-size"`
```
Comment (darioush) [inline plugin/evm/config.go:99]: should we avoid dropping the `tx-` prefix for backwards config compatibility for these 3 flags?

File: plugin/evm/gossiper.go [hunk 7]
```diff
@@ -185,15 +186,50 @@ func (n *pushGossiper) queueRegossipTxs() types.Transactions {
...
+	rgFrequency := n.config.RegossipFrequency
+	rgMaxTxs := n.config.RegossipMaxTxs
+	localQueued := n.queueExecutableTxs(state, tip.BaseFee(), localTxs, rgFrequency, rgMaxTxs, 1)
 	localCount := len(localQueued)
-	if localCount >= n.config.TxRegossipMaxSize {
+	if localCount >= rgMaxTxs {
 		return localQueued
 	}
-	remoteQueued := n.queueExecutableTxs(state, tip.BaseFee(), remoteTxs, n.config.TxRegossipMaxSize-localCount)
+	remoteQueued := n.queueExecutableTxs(state, tip.BaseFee(), remoteTxs, rgFrequency, rgMaxTxs-localCount, 1)
```
Comment (darioush) [inline plugin/evm/gossiper.go:196]: nit: would read nicer if we had `maxTxPerAccount := 1` and pass that or a comment

File: plugin/evm/gossiper.go [hunk 8]
```diff
@@ -84,86 +84,87 @@ func (vm *VM) newPushGossiper() Gossiper {
...
 	}
 	net.awaitEthTxGossip()
 	return net
 }
 
+// addrStatus used to track the metadata of addresses being queued for
+// regossip.
+type addrStatus struct {
+	nonce uint64
+	added int
```
Comment (darioush) [inline plugin/evm/gossiper.go:97]: nit: `txsAdded` or `numTxsAdded` or similar reads slightly better IMO

File: plugin/evm/config.go (hunk 5 above)
Comment (patrick-ogrady) [reply plugin/evm/config.go:103]: whoops!

File: plugin/evm/config.go (hunk 6 above)
Comment (patrick-ogrady) [reply plugin/evm/config.go:99]: I think it is fine...no one currently uses these

File: plugin/evm/gossiper.go (hunk 7 above)
Comment (patrick-ogrady) [reply plugin/evm/gossiper.go:196]: I'm just going to make this another config

File: plugin/evm/gossiper.go [hunk 9]
```diff
@@ -185,15 +186,51 @@ func (n *pushGossiper) queueRegossipTxs() types.Transactions {
...
-	remoteQueued := n.queueExecutableTxs(state, tip.BaseFee(), remoteTxs, n.config.TxRegossipMaxSize-localCount)
+	remoteQueued := n.queueExecutableTxs(state, tip.BaseFee(), remoteTxs, rgFrequency, rgMaxTxs-localCount, rgTxsPerAddr)
 	return append(localQueued, remoteQueued...)
 }
 
+// queueRegossipTxs finds the best priority transactions in the mempool and adds up to
+// [PriorityRegossipMaxTxs] of them to [txsToGossip].
+func (n *pushGossiper) queuePriorityRegossipTxs() types.Transactions {
+	// Fetch all pending transactions
+	pending := n.txPool.Pending(true)
```
Comment (ceyonur) [inline plugin/evm/gossiper.go:205]: Would be something like this in mempool useful here to not actually traverse all other accounts (in both for effective tip filtering in mempool and here to prune other addresses)?
```
func (pool *TxPool) PendingFrom(from common.Address, enforceTips bool) types.Transactions {
...
}
```

Comment (aaronbuchwald) [reply plugin/evm/gossiper.go:205]: Ya that's a good idea, I'll add that change in.

File: plugin/evm/gossiper.go [hunk 10]
```diff
@@ -84,86 +84,87 @@ func (vm *VM) newPushGossiper() Gossiper {
...
+// [RegossipMaxTxs] of them to [txsToGossip].
 func (n *pushGossiper) queueRegossipTxs() types.Transactions {
 	// Fetch all pending transactions
 	pending := n.txPool.Pending(true)
 
+	// Remove all priority transactions
+	for _, account := range n.config.PriorityRegossipAddresses {
+		delete(pending, account)
+	}
+
```
Comment (aaronbuchwald) [inline plugin/evm/gossiper.go:167]: Can we remove this and allow priority transactions to be included in both sets of gossip?

File: core/tx_pool.go [hunk 11]
```diff
@@ -606,6 +606,34 @@ func (pool *TxPool) Pending(enforceTips bool) map[common.Address]types.Transacti
 	return pending
 }
 
+func (pool *TxPool) PendingFrom(addrs []common.Address, enforceTips bool) map[common.Address]types.Transactions {
+	pool.mu.Lock()
+	defer pool.mu.Unlock()
+
+	pending := make(map[common.Address]types.Transactions)
+	for _, addr := range addrs {
+		list := pool.pending[addr]
```
[the remark this answers is not in the document]
Comment (darioush) [reply core/tx_pool.go:615]: +1 to this comment

File: core/tx_pool.go [hunk 12]
```diff
@@ -606,6 +606,34 @@ func (pool *TxPool) Pending(enforceTips bool) map[common.Address]types.Transacti
 	return pending
 }
 
+func (pool *TxPool) PendingFrom(addrs []common.Address, enforceTips bool) map[common.Address]types.Transactions {
```
Comment (darioush) [inline core/tx_pool.go:609]: can we add a comment for this function?

**Review by ceyonur — APPROVED** [review]

**Comment by darioush** [comment]
this is pretty cool, for the record.

**Review by aaronbuchwald — APPROVED** [review]
LGTM

**Pull request merged**